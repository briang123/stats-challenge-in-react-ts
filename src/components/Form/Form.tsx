import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { Button } from '../index';
import NumberInput from '../NumberInput/NumberInput';
import styled from 'styled-components';

interface IProps {
  setDataSet: (value: (number | string)[]) => void;
  dataSet: (number | string)[];
}

export const Form = ({ setDataSet, dataSet }: IProps) => {
  const [dataValue, setDataValue] = useState('');
  const inputRef = useRef<HTMLInputElement>(null);

  const handleSubmit = (e: React.FormEvent<HTMLButtonElement | HTMLInputElement>) => {
    e.preventDefault();
    setDataSet([...dataSet, dataValue]);
    setDataValue('');
    inputRef.current?.focus();
  };

  return (
    <FormContainer data-testid="form-container">
      <InputButtonGroup>
        <NumberInput
          testid="datapoint"
          ref={inputRef}
          tabIndex={0}
          placeholder="Enter a number"
          value={dataValue}
          autoFocus={true}
          setValue={setDataValue}
          onSubmit={handleSubmit}
        />
        <Button
          testid="submit-btn"
          onClick={handleSubmit}
          text="Submit"
          btnType="submit"
          disabled={!dataValue}
        />
      </InputButtonGroup>
    </FormContainer>
  );
};

Form.propTypes = {
  setDataSet: PropTypes.func.isRequired,
  dataSet: PropTypes.array.isRequired,
};

Form.defaultProps = {
  setDataSet: () => null,
  dataSet: [],
};

export default Form;

const FormContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  height: 200px;
  min-width: 100vw;
  align-items: center;
  justify-content: space-around;
  background-color: var(--secondarybg);
`;

const InputButtonGroup = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-around;
  min-width: 80vw;
  @media (min-width: 600px) {
    min-width: 40vw;
  }
`;
