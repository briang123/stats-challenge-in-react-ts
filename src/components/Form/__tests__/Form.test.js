import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';
import Form from '../Form';

afterEach(() => {
  cleanup();
  console.error.mockClear();
});

console.error = jest.fn();
const mockFn = jest.fn();

describe('<Form />', () => {
  it('Form component without props should not log error', () => {
    render(<Form />);
    expect(console.error).not.toHaveBeenCalled();
  });

  it('Form component without props should have an input group container with an input and button', () => {
    const { getByTestId } = render(<Form />);

    const container = getByTestId('form-container');
    expect(container.childElementCount).toBe(1);

    const group = container.firstChild;
    expect(group.childElementCount).toBe(2);

    const input = getByTestId('datapoint');
    expect(input).not.toBeNull();

    const button = getByTestId('submit-btn');
    expect(button).not.toBeNull();
  });

  it('Form component without props should render input with placeholder "Enter a value"', () => {
    const { getByTestId } = render(<Form />);

    const input = getByTestId('datapoint');
    expect(input.getAttribute('placeholder')).toEqual('Enter a number');
  });

  it('Form component without props should render input of type number', () => {
    const { getByTestId } = render(<Form />);

    const input = getByTestId('datapoint');
    expect(input.getAttribute('type')).toBe('number');
  });

  it('Form component without props should render input with null value', () => {
    const { getByTestId } = render(<Form />);

    const input = getByTestId('datapoint');
    expect(input.nodeValue).toBeNull();
  });

  it('Form component without props should render input with tabindex set to 0', () => {
    const { getByTestId } = render(<Form />);

    const input = getByTestId('datapoint');
    expect(input.getAttribute('tabindex')).toBe('0');
  });

  it('Form component without props should render submit button with tabindex = -1', () => {
    const { getByTestId } = render(<Form />);

    const button = getByTestId('submit-btn');
    expect(button.getAttribute('tabindex')).toBe('-1');
  });

  it('Form component without props should render the submit button disabled', () => {
    const { getByTestId } = render(<Form />);

    const button = getByTestId('submit-btn');
    expect(button.getAttribute('disabled')).toBe('');
  });

  it('Form component without props should render submit button with button text "Submit"', () => {
    const { getByTestId } = render(<Form />);

    const button = getByTestId('submit-btn');
    expect(button.textContent).toBe('Submit');
  });

  it('Form component with props should have an input group container with an input and button', () => {
    const { getByTestId } = render(
      <Form data-testid="form" dataSet={[1, 2]} setDataSet={mockFn} />
    );

    const container = getByTestId('form-container');
    expect(container.childElementCount).toBe(1);

    const group = container.firstChild;
    expect(group.childElementCount).toBe(2);

    const input = getByTestId('datapoint');
    expect(input).not.toBeNull();

    const button = getByTestId('submit-btn');
    expect(button).not.toBeNull();
  });

  it('Input should accept only numbers', () => {
    const { getByTestId } = render(
      <Form data-testid="form" dataSet={[1, 2]} setDataSet={mockFn} />
    );

    const input = getByTestId('datapoint');

    expect(input.value).toBe('');
    fireEvent.change(input, { target: { value: 'foo' } });
    expect(input.value).toBe('');
  });

  it('Submit button should be enabled when user types a number in the input', () => {
    const { getByTestId } = render(
      <Form data-testid="form" dataSet={[1, 2]} setDataSet={mockFn} />
    );

    const input = getByTestId('datapoint');
    fireEvent.change(input, { target: { value: '1' } });
    expect(input.value).toBe('1');

    const button = getByTestId('submit-btn');
    expect(button.getAttribute('disabled')).toBeNull();
  });

  it('Should match snapshot with props', () => {
    const tree = create(
      <Form data-testid="form" dataSet={[1, 2]} setDataSet={mockFn} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
