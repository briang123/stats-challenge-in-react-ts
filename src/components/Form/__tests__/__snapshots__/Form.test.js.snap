// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> Should match snapshot with props 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  border: none;
  height: 4.87rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: var(--lightcolor);
  outline: none;
  width: 30%;
  border-radius: 0px 20px 20px 0px;
  background-color: var(--primarybtn);
}

.c3:disabled {
  background-color: var(--disabledbtn);
}

.c3:hover {
  cursor: pointer;
}

.c2 {
  height: 75px;
  width: 70%;
  border-radius: 20px 0px 0px 20px;
  border: none;
  font-size: calc(20px + 2vmin);
  text-align: center;
  -moz-appearance: textfield;
}

.c2:focus {
  outline: none;
}

.c2::-webkit-input-placeholder {
  color: var(--disabledbtn);
}

.c2::-moz-placeholder {
  color: var(--disabledbtn);
}

.c2:-ms-input-placeholder {
  color: var(--disabledbtn);
}

.c2:-moz-placeholder {
  color: var(--disabledbtn);
}

.c2::-webkit-outer-spin-button,
.c2::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  height: 200px;
  min-width: 100vw;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  background-color: var(--secondarybg);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  min-width: 80vw;
}

@media (min-width:600px) {
  .c1 {
    min-width: 40vw;
  }
}

<div
  className="c0"
  data-testid="form-container"
>
  <div
    className="c1"
  >
    <input
      autoFocus={true}
      className="c2"
      data-testid="datapoint"
      onChange={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      placeholder="Enter a number"
      tabIndex={0}
      type="number"
      value=""
    />
    <button
      className="c3"
      data-testid="submit-btn"
      disabled={true}
      onClick={[Function]}
      tabIndex={-1}
    >
      Submit
    </button>
  </div>
</div>
`;
